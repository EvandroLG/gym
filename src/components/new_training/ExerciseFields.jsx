import React, { Component } from 'react';

export default class ExerciseFields extends Component {

  constructor(props) {
    super(props);

    this.state = {
      name: '',
      repetition: '',
      weight: '',
      set: '',
      youtube: ''
    };
  }

  componentDidMount() {
    if (this.nameInput) {
      this.nameInput.focus();
    }
  }

  onInputChange(e) {
    const key = e.target.id.split('_')[1];

    this.setState({
      [key]: e.target.value
    }, () => {
      this.props.onInputChange(this.props.index + '', this.state)
    });
  }

  render() {
    return (
      <div className="row form-group mt-1 d-flex">
        <div className="col">
          <label htmlFor={`exercise_name_${this.props.index}`}>Name</label>

          <input type="text" className="form-control"
            ref={ (input) => this.nameInput = input }
            id={`exercise_name_${this.props.index}`}
            onChange={this.onInputChange.bind(this)} />
        </div>

        <div className="col">
          <label htmlFor={`exercise_set_${this.props.index}`}>Set</label>

          <input type="text" className="form-control"
            id={`exercise_set_${this.props.index}`}
            onChange={this.onInputChange.bind(this)} />
        </div>

        <div className="col">
          <label htmlFor={`exercise_repetition_${this.props.index}`}>Repetition</label>

          <input type="text" className="form-control"
            id={`exercise_repetition_${this.props.index}`}
            onChange={this.onInputChange.bind(this)} />
        </div>

        <div className="col">
          <label htmlFor={`exercise_weight_${this.props.index}`}>Weight</label>

          <input type="text" className="form-control"
            id={`exercise_weight_${this.props.index}`}
            onChange={this.onInputChange.bind(this)} />
        </div>

        <div className="col">
          <label htmlFor={`exercise_youtube_${this.props.index}`}>Youtube</label>

          <input type="text" className="form-control"
            id={`exercise_youtube_${this.props.index}`}
            onChange={this.onInputChange.bind(this)} />
        </div>

        <div className="col align-self-center">
          <button type="button" className="btn btn-danger btn-sm"
            onClick={ () => this.props.onButtonRemoveExercise(this.props.index) }>
            Remove
          </button>
        </div>
      </div>
    )
  }
}
