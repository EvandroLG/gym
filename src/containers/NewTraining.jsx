import { connect } from 'react-redux';
import { setTitleTraining } from '../action';
import NewTraining from '../components/new_training/Index';

const mapStateToProps = ({ newTraining }) => {
  return {
    title: newTraining.title
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onTitleUpdate: (value) => {
      dispatch(setTitleTraining(value));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(NewTraining);
