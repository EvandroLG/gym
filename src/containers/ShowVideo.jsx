import { connect } from 'react-redux';
import { setVideoUrl } from '../action';
import ShowVideo from '../components/training_list/ShowVideo';

export const mapStateToProps = ({ videoUrl }) => {
  return {
    videoUrl
  };
};

export const mapDispatchToProps = (dispatch) => {
  return {
    onVideoUrlChange: (url) => {
      dispatch(setVideoUrl(url));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ShowVideo);
