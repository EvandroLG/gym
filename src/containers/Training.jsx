import { connect } from 'react-redux';
import { setVideoUrl } from '../action';
import Training from '../components/training_list/Training'

const mapStateToProps = ({ videoUrl }) => {
  return {
    videoUrl
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onVideoUrlChange: (url) => {
      dispatch(setVideoUrl(url));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Training);
