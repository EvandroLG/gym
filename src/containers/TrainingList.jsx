import { connect } from 'react-redux';
import DB from '../libs/db';
import TrainingList from '../components/training_list/';

const mapStateToProps = ({ trainingList }) => {
  return {
    trainingList
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    findAll: () => {
      new DB().findAll((trainingList) => {
        dispatch({
          type: "LOAD_TRAINING_LIST",
          trainingList,
        });
      });
    },

    onRemoveTraining: (id) => {
      new DB().remove(id);

      dispatch({
        type: 'REMOVE_TRAINING',
        id
      });
    },

    onAddNewExercise: (id) => {
      dispatch({
        type: 'ADD_NEW_EXERCISE',
        id
      });
    },

    onRemoveExercise: (idTraining, idExercise) => {
      dispatch({
        type: 'REMOVE_EXERCISE',
        idTraining,
        idExercise
      });
    },

    onChangeTitle: (id, title) => {
      dispatch({
        type: 'CHANGE_TITLE_TRAINING',
        id,
        title
      });
    },

    onChangeExercise: (idTraining, idExercise, property, value) => {
      dispatch({
        type: 'CHANGE_EXERCISE_TRAINING',
        idTraining,
        idExercise,
        property,
        value
      });
    },

    onSubmit: (id, training) => {
      new DB().update(id, training);
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(TrainingList);
