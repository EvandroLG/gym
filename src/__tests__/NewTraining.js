import React from 'react';
import { shallow, mount } from 'enzyme';
import fixtures from './fixtures';
import DB from '../libs/db';
import NewTraining from '../components/new_training/Index';
import ExerciseFields from '../components/new_training/ExerciseFields';
import { mapStateToProps as mapStateNewTraining, mapDispatchToProps } from '../containers/NewTraining';
jest.mock('../libs/db');

/* global verifySnapshot */

describe('container', () => {
  it('should contains title and exerciseList with correct values in mapStateToProps', () => {
    const { title, exerciseList } = fixtures[0];

    verifyMapStateToProps(mapStateNewTraining, {
      newTraining: { title }
    }, 'title', title);

    verifyMapStateToProps(mapStateNewTraining, {
      newTraining: { exerciseList }
    }, 'exerciseList',  exerciseList);
  });

  it('should return onTitleUpdate method', () => {
    const title = 'Arms';

    verifyMapDispatchToProps(mapDispatchToProps, 'onTitleUpdate', [title], {
      type: 'SET_TITLE_NEW_TRAINING',
      title
    });
  });

  it('should return onAddExercise method', () => {
    verifyMapDispatchToProps(mapDispatchToProps, 'onAddExercise', [], {
      type: 'ADD_EXERCISE_NEW_TRAINING'
    });
  });

  it('should return onRemoveExercise', () => {
    const id = 1;
    verifyMapDispatchToProps(mapDispatchToProps, 'onRemoveExercise', [id], {
      type: 'REMOVE_EXERCISE_NEW_TRAINING',
      id
    });
  });

  it('should return onUpdateExercise', () => {
    const id = 1;
    const fieldName = 'name';
    const value = 'Pull-Down';

    verifyMapDispatchToProps(mapDispatchToProps, 'onUpdateExercise', [id, fieldName, value], {
      type: 'SET_FIELD_NEW_TRAINING',
      id,
      fieldName,
      value
    });
  });

  it('it should return onSubmit', () => {
    const mockCallback = jest.fn();
    const { title, exerciseList } = fixtures[0];

    mapDispatchToProps(mockCallback).onSubmit(title, exerciseList);

    expect(DB.prototype.insert).toBeCalled();
    expect(DB.prototype.insert.mock.calls[0][0]).toEqual({
      title,
      exerciseList
    });

    DB.prototype.insert.mock.calls[0][1]();

    expect(mockCallback).toBeCalled();
    expect(mockCallback).toBeCalledWith({
      type: 'EMPTY_NEW_TRAINING'
    });
  });
});

describe('components', () => {

  describe('new training component', () => {
    function mountNewTraining(keyProp, mockCallback) {
      const props = {
        title: '',
        exerciseList: [],
        [keyProp]: mockCallback
      };

      return mount(<NewTraining { ...props } />);
    }

    it('new training component should render as expected', () => {
      const newTraining = mountNewTraining();
      verifySnapshot(newTraining);
    });

    it('should call correct callback after click on add button', () => {
      const mockCallback = jest.fn();
      const newTraining = mountNewTraining('onAddExercise', mockCallback) 

      newTraining.find('#add').simulate('click');
      expect(mockCallback).toBeCalled();
    });

    it('should call correct callback after change title', () => {
      const mockCallback = jest.fn();
      const newTraining = mountNewTraining('onTitleUpdate', mockCallback);

      newTraining.find('#title').simulate('change', {
        target: {
          value: 'Chest'
        }
      });

      expect(mockCallback).toMatchSnapshot();
    });

    it('should call correct method after submit form', () => {
      const { title, exerciseList } = fixtures[0];
      const mockCallback = jest.fn();

      const props = {
        title,
        exerciseList,
        onSubmit: mockCallback
      };

      const newTraining = mount(<NewTraining { ...props } />);

      newTraining.find('form').simulate('submit', {
        preventDefault: () => {}
      });

      expect(mockCallback).toBeCalled();
      expect(mockCallback).toBeCalledWith(title, exerciseList);
    });
  });

  describe('exercise fields component', () => {
    function mountExerciseFields(keyProperty, mockCallback) {
      const props = {
        id: 1,
        [keyProperty]: mockCallback
      };

      return shallow(<ExerciseFields { ...props } />);
    }

    it('exercise fields component should render as expected', () => {
      const exerciseFields = mountExerciseFields();
      verifySnapshot(exerciseFields);
    });

    it('should call correct method after change input', () => {
      const mockCallback = jest.fn();
      const exerciseFields = mountExerciseFields('onInputChange', mockCallback);
      const id = 'exercise_name_1';
      const value = 'Pull-Down';

      exerciseFields.find(`#${id}`).simulate('change', {
        target: {
          id,
          value
        }
      });

      expect(mockCallback).toBeCalled();
      expect(mockCallback).toBeCalledWith(1, 'name', value);
    });

    it('should call correct callback after click on remove button', () => {
      const mockCallback = jest.fn();
      const exercise = mountExerciseFields('onButtonRemoveExercise', mockCallback);

      exercise.find('.remove').simulate('click');

      expect(mockCallback).toMatchSnapshot();
    });
  });
});
