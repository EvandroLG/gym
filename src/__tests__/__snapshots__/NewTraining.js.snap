// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components exercise fields component exercise fields component should render as expected 1`] = `
<div
  className="row"
>
  <div
    className="col"
  >
    <label
      htmlFor="exercise_name_1"
    >
      Name
    </label>
    <input
      id="exercise_name_1"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="col"
  >
    <label
      htmlFor="exercise_set_1"
    >
      Set
    </label>
    <input
      id="exercise_set_1"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="col"
  >
    <label
      htmlFor="exercise_repetition_1"
    >
      Repetition
    </label>
    <input
      id="exercise_repetition_1"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="col"
  >
    <label
      htmlFor="exercise_weight_1"
    >
      Weight
    </label>
    <input
      id="exercise_weight_1"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="col"
  >
    <label
      htmlFor="exercise_youtube_1"
    >
      Youtube
    </label>
    <input
      id="exercise_youtube_1"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="col"
  >
    <button
      className="remove"
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
</div>
`;

exports[`components exercise fields component should call correct callback after click on remove button 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      1,
    ],
  ],
}
`;

exports[`components new training component new training component should render as expected 1`] = `
<NewTraining
  exerciseList={Array []}
  title=""
>
  <div>
    <h1>
      Register a new training
    </h1>
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          id="title"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          className="add"
          id="add"
          type="button"
        >
          Add Exercise
        </button>
      </div>
      <input
        type="submit"
        value="Register it!"
      />
    </form>
  </div>
</NewTraining>
`;

exports[`components new training component should call correct callback after change title 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Chest",
    ],
  ],
}
`;
